src/modsim2/similarity/metrics/kde.py

```diff
@@ -0,0 +1,295 @@
+    A metric is then applied to calculate a distance between the
+    density distributions.
+
+    sklearn's KernelDensity class has been used as a variety of
+    kernels are implemented and the bandwidth can be estimated
+    using Scott's Rule.
+    """
+
+    def __init__(self, seed: int) -> None:
+        super().__init__(seed)
+
+    def _pre_process_data(self, data_A: np.ndarray, data_B: np.ndarray):
+        pass
+
+    @staticmethod
+    def l2(
+        num_dimensions: int,
+        estimator_A: KernelDensity,
+        estimator_B: KernelDensity,
+        integration_kwargs: dict,
+    ) -> float:
+        """
+        Calculates the L2 Norm between two probability density estimatiors for a given
+        number of dimensions using integration
+
+        Args:
+            num_dimensions : the number of dimensions that are to be integrated over
+            estimator_A: the density estimator that has been generated for a dataset A
+            estimator_B: the density estimator that has been generated for a dataset B
+
+        Returns:
+            distance: the L2 distance
+        """
+        # the function to be integrated
+        def func(*args):
+            return np.power(
+                np.exp(estimator_A.score([[*args]]))
+                - np.exp(estimator_B.score([[*args]])),
+                2,
+            )
+
+        # the bounds over which the integration will be performed
+        bounds = [[0, 1] for _ in range(num_dimensions)]
+
+        # perform integration
+        distance = integrate.nquad(func, bounds, **integration_kwargs)[0]
+        distance = np.sqrt(distance)
+
+        return distance
+
+    @staticmethod
+    def kl(
+        num_dimensions: int,
+        estimator_A: KernelDensity,
+        estimator_B: KernelDensity,
+        integration_kwargs: dict,
+    ):
+        """
+        Calculates the KL divergence between two probability density estimatiors for a
+        given number of dimensions using integration
+
+        Args:
+            num_dimensions : the number of dimensions that are to be integrated over
+            estimator_A: the density estimator that has been generated for a dataset A
+            estimator_B: the density estimator that has been generated for a dataset B
+
+        Returns:
+            divergence: the KL divergence
+        """
+        # the function to be integrated
+        def func(*args):
+            return np.exp(estimator_A.score(np.array([[*args]]))) * (
+                np.log(
+                    (np.exp(estimator_A.score([[*args]])))
+                    / (np.exp(estimator_B.score(np.array([[*args]]))))
+                )
+            )
+
+        # the bounds over which the integration will be performed
+        bounds = [[0, 1] for _ in range(num_dimensions)]
+
+        # perform integration
+        divergence = integrate.nquad(func, bounds, **integration_kwargs)[0]
+
+        return divergence
+
+    @staticmethod
+    def tv(
```

Comment:
rename `total_variation` unless tv is common usage in literature?

Reply:
done